### ISSUE TITLE
Incorrect warning CS0649 (field never assigned) with ref return

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn#10544]
**Version Used**: Visual Studio "15" Preview

**Steps to Reproduce**:
1. Write the following class in a project with the demo features enabled (`__DEMO__` compilation symbol):
   
   ```
   class Test
   {
       private int x;
       public ref int GetRefOfX()
       {
           return ref x;
       }
   }
   ```
2. Build

**Expected Behavior**:

It should compile without warning.

**Actual Behavior**:

The compiler produces warning CS0649:

> Field 'Program.Test.x' is never assigned to, and will always have its default value 0

This warning is incorrect, because even though `x` isn't explicitly initialized in the class, a ref to it can be returned, so it can be assigned by a caller, e.g.:

```
var t = new Test();
t.GetRefOfX() = 42;
```

**Comment by HaloFour** [comment]
I think the warning still makes sense.  Even though the caller *could* assign to the ref, they could just as easily just dereference the ref and get that "uninitialized" default value.  You can eliminate the warning by simply assigning a value to the field and being more explicit about what you're doing.

**Comment by thomaslevesque** [comment]
@HaloFour, I see your point, but I still think the message is confusing; the compiler shouldn't make assumptions about like _will always have its default value 0_, especially when it knows that we return a reference to it.

**Comment by HaloFour** [comment]
@thomaslevesque 

The compiler knows that the field will have its default value of 0 due to the heap memory being zero-initialized.  You get the exact same warning without using `ref` returns for the same reason; at the point that you're using the value of `x` the compiler has detected that you've never assigned a value to it.

```cs
class Test
{
    private int x;
    public int GetX()
    {
        return x; // CS0649
    }
}
```

**Comment by thomaslevesque** [comment]
@HaloFour, that's a different issue. Returning `x` by value doesn't let caller change the value of `x`, so the warning is normal in this case. Having the same message when `x` is returned by reference is misleading, since there is a way for the value of `x` to be changed.

Perhaps the warning should remain, but at least the message should be different in this case.

**Comment by VSadov** [comment]
This is the same scenario as if a field was passed as a byref parameter. We should not report a warning

```C#
    //no warnings here
    class C1
    {
        internal int x;  

        public void M2()
        {
            M2(ref x);
        }

        private void M2(ref int arg)
        {

        }
    }
```

The warning is mostly to note the uselesness of a field that is never assigned to. But the analysis is pretty simple and catches only absolutely trivial cases. 
In the repro scenario, since the field has escaped via a ref, and we do not track usage through references, we no longer can claim it is never assigned.

Thanks for finding this!!

**Comment by jaredpar** [comment]
@VSadov is this fixed by the referenced PRs?

**Comment by VSadov** [comment]
fixed by https://github.com/dotnet/roslyn/pull/14418
